20251028-BRS-workflow-versioning.md

```diff
@@ -0,0 +1,416 @@
+effort by existing Dapr developers.
+
+## Implementation Details
+I propose that the internal Dapr Workflow API be slightly modified to include a string value whenever a workflow is
+executed. This value contains the name of the versioned type that ran for any given workflow invocation as well as a 
+list of each named patch applied and run in that execution.
+
+Different Workflow SDKs should introduce a variety of versioning strategies that best accommodate their unique
+capabilities and allow the user to opt-into versioning and select a strategy at startup. Documentation should be written
+that strongly urges developers to pick one strategy and stick with it, though future iterations of this in the SDKs
+could certainly introduce some mechanism to migrate from one strategy to another.
+
+Workflows and activities will be registered with the Dapr Workflow SDK as they are today.
+
+Whenever a request come into the application from the Dapr runtime to run work on a specific workflow following a 
+`ScheduleNewWorkflowAsync` call, during the same following a `ContinueAsNew` call or even following a 
+`CallChildWorkflowAsync`, a request should come into the SDKs to run that named workflow.
+
+### Version isn't specified on orchestration start event
+If there is no version specified on the event, the SDK should map the request based on the named base type to the
+latest version available in the application and run the workflow there. Every time a patch is evaluated, the "new code" 
+route should be taken, and the patch identifier persisted in a list of encountered patches.
+
+At each boundary invocation except workflow orchestration completion, the SDK should return the workflow status with
+the (potentially updated) version string defined below containing the name of the workflow type executed and the list 
+of named patches followed.
+
+### Version is specified on orchestration start event
+If the version is supplied, the workflow type should be parsed out and the SDK should ignore "latest" evaluations and use
+that type to run the workflow execution. The list of patches should be extracted from the version string. At each
+evaluation of the patch, the path should be taken that corresponds to whether the patch exists in the list:
+
+- If the patch exists in the version list, the "new code" route should be taken
+- If the patch does not exist in the version list, the "old code" route should be taken
+
+#### Version string format
+The version string should use double pipe characters to separate the different parts of the value. It should always start
+with the name of the workflow type, followed by a mandatory '||' and then a list of each case-sensitive named patch applied 
+and used in the execution surrounded with square brackets and themselves separated by double pipe characters. The brackets 
+should always be present whether there are any patches named within them or not. An example without any patches might 
+look like:
```

Comment:
Durabletask is defined in proto with fully extensible messages, there shouldn't be any need to overload a single field (workflow name) with more data. The version should be represented as a new optional field.

More this- older clients will break entirely after the introduction of this new schema.